testsuite/MDAnalysisTests/analysis/test_contacts.py

```diff
@@ -171,16 +171,64 @@ class TestContacts(object):
     def universe():
         return mda.Universe(PSF, DCD)
 
-    def _run_Contacts(self, universe, start=None, stop=None, step=None, **kwargs):
+    def _run_Contacts(self, universe, sel_acidic=None, sel_basic=None, start=None, stop=None, step=None, **kwargs):
         acidic = universe.select_atoms(self.sel_acidic)
         basic = universe.select_atoms(self.sel_basic)
+        if sel_acidic is None:
+            sel_acidic = self.sel_acidic
+        if sel_basic is None:
+            sel_basic = self.sel_basic
+
         return contacts.Contacts(
             universe,
-            select=(self.sel_acidic, self.sel_basic),
+            select=(sel_acidic, sel_basic),
             refgroup=(acidic, basic),
             radius=6.0,
             **kwargs).run(start=start, stop=stop, step=step)
 
+    def test_select_types(self, universe):
+        """Test if Contact can take both string and AtomGroup as selections.
+        """
+        acidic = universe.select_atoms(self.sel_acidic)
+        basic = universe.select_atoms(self.sel_basic)
+
+        with_string = self._run_Contacts(universe)
+        with_atomgroup = self._run_Contacts(universe, sel_acidic=acidic, sel_basic=basic)
+       
+        assert_equal(with_atomgroup.grA, with_string.grA)
+        assert_equal(with_atomgroup.grB, with_string.grB) 
+
+    @pytest.mark.xfail
+    def test_select_updating_acid(self, universe):
+        uag = universe.select_atoms(self.sel_acidic, updating=True)
+        self._run_Contacts(universe, sel_acidic=uag)
+    
+    @pytest.mark.xfail
+    def test_select_updating_basic(self, universe):
+        uag = universe.select_atoms(self.sel_basic, updating=True)
+        self._run_Contacts(universe, sel_basic=uag)
+
+    @pytest.mark.xfail
+    def test_select_wrong_type(self, universe):
+        """test that Contacts fails if selection types are wrong"""
+        selections = [
+            (1, 2), 
+            ([1], [2]), 
+            (mda.Universe, mda.Universe),
+            (AtomGroup, 2)
+        ]
+        fails = True
+        for sel_acid, self_basic in selections:
+            try:
+                self._run_Contacts(universe, sel_acid=sel_acid, sel_basic=sel_basic)
+                fails = False
+                break
+            except:
+                pass
+        if fails:
+            raise TypeError
```

Comment:
Perhaps a bit awkward to catch all/generic `Exception`s in a test and manually raise a `TypeError` inside the test. Perhaps following the guidance related to `pytest.raises()` will clear that up to test for the expected errors in the code proper, since manually raising an error in the testsuite isn't of much use to end users of course.

If some of the parameters fail and others succeed you can xfail on a case-by-case basis as well.

Reply:
Thanks for you comment! I updated the code to use `@pytest.mark.parametrization` decorator, and I believe that should bring it in line with the style of `contacts.py`. I'm a bit uncomfortable with how I had to use the decorator, but maybe that's okay? Should we resolve?